### PR TITLE
Add draft for EOF with jumpdest table

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

**Comment by axic** [comment]
Editors: no need to review yet, we may want to push a few more commits to this.

**Review by VARsIsYoutacRS — APPROVED** [review]

File: EIPS/eip-3690.md [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+status: Draft
+type: Standards Track
+category: Core
+created: 2021-06-23
+requires: 3540, 3670
+---
+
+## Abstract
+
+Introduce a section in the EOF format ([EIP-3540](./eip-3540.md)) for storing the list of `JUMPDEST`s, validate the correctness of this list at the time of contract creation, and remove the need for `JUMPDEST`-analysis at execution time. The `JUMPDEST` instruction is not needed anymore and becomes invalid.
```
Comment (MicahZoltu) [inline EIPS/eip-3690.md:15]: May want to make it clear in the abstract that the JUMPDEST instruction is invalid in contracts of the new type, but they will remain valid in contracts of the old type.

Comment (MicahZoltu) [reply EIPS/eip-3690.md:15]: Recommend mentioning what the prefix for this new type of contract is in the abstract.

File: EIPS/eip-3690.md [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+author: Alex Beregszaszi (@axic), Paweł Bylica (@chfast), Andrei Maiboroda (@gumb0)
+discussions-to: https://ethereum-magicians.org/t/evm-object-format-eof/5727
+status: Draft
+type: Standards Track
+category: Core
+created: 2021-06-23
+requires: 3540, 3670
+---
+
+## Abstract
```
Comment (MicahZoltu) [inline EIPS/eip-3690.md:13]: I know this is going to fall on deaf ears, but I recommend a Simple Summary.
```suggestion
## Simple Summary
Adds a new type of contract that has section for storing jump destinations.

## Abstract
```

File: EIPS/eip-3690.md [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    # Ensure last PUSH doesn't go over code end
+    assert(pos == len(code))
+
+    # The table is invalid if there are untouched locations
+    assert(len(jumpdests) == 0)
+```
+
+## Test Cases
+
+#### Valid bytecodes
```
Comment (MicahZoltu) [inline EIPS/eip-3690.md:227]:
```suggestion
### Valid bytecodes
```
Skipping section levels can have weird affects on tooling (like the ToC generator).

File: EIPS/eip-3690.md [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+- No jumpdests
+- Every byte is a jumpdest
+- Distant jumpdests (0x7f and 0x3f01 bytes apart)
+- Max number of jumpdests
+    - 1-byte offset encoding: initcode of max size (64K) with jumpdest at each byte - table contains 65536 1-byte offsets, first one is 0x00, all others equal 0x01
+    - 2-byte offset encoding: inicode of max size with jumpdests 0x80 (128) bytes apart - table contains 512 offsets, first one is 0x7f (127), all others equal 0x8001
+    - 3-byte offset encoding: inicode of max size with jumpdests 0x4000 (16384) bytes apart - table contains 4 offsets: 0xFF7F (16383), 0x808001, 0x808001, 0x808001
+
+#### Invalid bytecodes
```
Comment (MicahZoltu) [inline EIPS/eip-3690.md:237]:
```suggestion
### Invalid bytecodes
```

**Review by MicahZoltu — APPROVED** [review]
Some feedback, but I don't think any blockers (because it has been decided that I am not allowed to block on missing Simple Summary, otherwise I would block on that).

File: EIPS/eip-3690.md (hunk 1 above)
Comment (axic) [reply EIPS/eip-3690.md]: > Recommend mentioning what the prefix for this new type of contract is in the abstract.

We don't have that nailed down yet, but will rework this at the same time we rework the other EIPs (soon).

Comment (axic) [reply EIPS/eip-3690.md]: > May want to make it clear in the abstract that the JUMPDEST instruction is invalid in contracts of the new type, but they will remain valid in contracts of the old type.

Added this.

File: EIPS/eip-3690.md (hunk 3 above)
Comment (axic) [reply EIPS/eip-3690.md:226]: Thanks, fixed.

File: EIPS/eip-3690.md (hunk 2 above)
Comment (axic) [reply EIPS/eip-3690.md:13]: Thanks.  I think we'll have to consider that as the abstract is growing, but I'd opt that we do that at once for all the relevant EIPs (need more discussions to come up concise descriptions for all).

If it is not a blocker, I'd opt for merging in the current state.

**Comment by axic** [comment]
@MicahZoltu made the changes and left a comment regarding the simple summary.

**Comment by MicahZoltu** [comment]
Still approved then.  😉

**Pull request merged**